packages/ember-application/lib/system/application-instance.js

```diff
@@ -212,6 +298,255 @@ let ApplicationInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
   }
 });
 
+if (isEnabled('ember-application-visit')) {
+  ApplicationInstance.reopen({
+    /**
+      Returns the current URL of the app instance. This is useful when your
+      app does not update the browsers URL bar (i.e. it uses the `'none'`
+      location adapter).
+
+      @public
+      @return {String} the current URL
+    */
+    getURL() {
+      let router = get(this, 'router');
+      return get(router, 'url');
+    },
+
+    // `instance.visit(url)` should eventually replace `instance.handleURL()`;
+    // the test helpers can probably be switched to use this implementation too
+
+    /**
+      Navigate the instance to a particular URL. This is useful in tests, for
+      example, or to tell the app to start at a particular URL. This method
+      returns a promise that resolves with the app instance when the transition
+      is complete, or rejects if the transion was aborted due to an error.
+
+      @public
+      @param url {String} the destination URL
+      @return {Promise}
+    */
+    visit(url) {
+      this.setupRouter();
+
+      let router = get(this, 'router');
+
+      let handleResolve = () => {
+        // Resolve only after rendering is complete
+        return new Ember.RSVP.Promise((resolve) => {
+          // TODO: why is this necessary? Shouldn't 'actions' queue be enough?
+          // Also, aren't proimses supposed to be async anyway?
```

Comment:
(But because this implementation is doing a correct thing, it is also not super urgent/blocker for merge)